evaluator/src/main/kotlin/ProjectSourceRule.kt

```diff
@@ -38,6 +38,15 @@ open class ProjectSourceRule(
      */
     val projectSourcesDir: File by lazy { projectSourceResolver.rootDir }
 
+    /**
+     * Return all directories from the project's source tree which match any of the
+     * provided [glob expressions][patterns].
+     */
+    fun projectSourceFindDirectories(vararg patterns: String): List<File> =
+        projectSourcesDir.walkBottomUp().filterTo(mutableListOf()) {
+            it.isDirectory && FileMatcher.match(patterns.toList(), it.relativeTo(projectSourcesDir).path)
+        }
```

Comment:
Unless the function is used in a different context in later commits it is inefficient to scan the whole file tree if the first match already gives the required result.

Reply:
It is used in a later commit. It is pubic so that it can be used from the user's `rules.kt` file, either directly from the rule or for composing a new matcher. Can we leave this performance optimization out of scope here?